Add Spl\ForwardArrayIterator and Spl\ReverseArrayIterator

morrisonlevi opened the pull request:
## Spl\ReverseArrayIterator

`Spl\ReverseArrayIterator` iterates over the array in reverse order, preserving keys.

```php
$array = [ 
    "one" => 1,
    "two" => 2,
    "three" => 3,
];

$iterator = new Spl\ReverseArrayIterator($array);
foreach ($iterator as $key => $current) {
    echo "[{$key} => {$current}]\n";
}
/* outputs:
[three => 3]
[two => 2]
[one => 1]
*/
```

## Spl\ForwardArrayIterator

ForwardArrayIterator is a simple array iterator which does not dup the array (though if the user modifies it after making an iterator then normal copy-on-write will apply and make a copy then). It iterates over the array in order. The API is intentionally very small. In contrast, `ArrayIterator` has a large API which requires it to almost always duplicate the array.

`Spl\ForwardArrayIterator` is intended to replace many usages of `new \ArrayIterator($array)` where iterating over the array in an iterator form is the intended use case. It is not intended to replace other usages, such as `ArrayIterator::sort`.

```php
$array = [ 
    "one" => 1,
    "two" => 2,
    "three" => 3,
];

$iterator = new Spl\ForwardArrayIterator($array);
foreach ($iterator as $key => $current) {
    echo "[{$key} => {$current}]\n";
}
/* outputs:
[one => 1]
[two => 2]
[three => 3]
*/
```

For such simple cases I recommend using the `$array` directly, but sometimes an iterator is needed for generic code, such as:

```php

function reduce_values(iterable $of, callable $by)
{
    // Uses Iterator API to avoid re-implementing it for arrays 
    $iterator = to_iterator($of);
    $iterator->rewind();
    if (!$iterator->valid()) {
        throw new \ValueException(
            __FUNCTION__ . " requires an iterable that contains at least one value"
        );
    }
    $result = $iterator->current();
    for ($iterator->next(); $iterator->valid(); $iterator->next()) {
        $result = $by($result, $iterator->current());
    }
    return $result;
}

// to_iterator may look something like this:
function to_iterator(iterable $subject): \Iterator
{
    if (\is_array($subject)) {
        /* Use \Spl\ForwardArrayIterator over \ArrayIterator so
         * so that we don't duplicate the $subject just to
         * construct an iterator.
         */
        return new \Spl\ForwardArrayIterator($subject);
    } elseif ($subject instanceof \Iterator) {
        return $subject;
    } else /* if ($subject instanceof \IteratorAggregate) */ {
        /* Technically IteratorAggregate::getIterator doesn't
         * have to return an Iterator, just any Traversable.
         * Protect it with a recursive call.
         */
        return namespace\to_iterator($subject->getIterator());
    }
}
```

## Status

  - [x] Spl\ForwardArrayIterator
      - [x] Countable
      - [x] Iterator
      - [x] Forbid unserialization and serialization
      - [x] Tests
  - [x] Spl\ReverseArrayIterator
      - [x] Countable
      - [x] Iterator
      - [x] Forbid unserialization and serialization
      - [x] Tests

TysonAndre commented:
Some notes on the implementation:

- ReflectionProperty->setValue(), `unserialize()`, etc. may cause unexpected changes in the properties being iterated over, modify the array or fields on the array. My preference would be to override the property array getters like SplObjectStorage currently does, and provide `getInternalOffset(): int` or maybe just `getArraySliceBefore(): array`/`getArraySliceAfter(): array` or `countElementsBefore()` or possibly nothing. e.g. to deal with gaps such as `<?php $x = ['a','b','c']; unset($x[1]); new Spl\ForwardArrayIterator($x)`.

  Closure::bindTo() is also a concern for modifying $this->offset
- Future scope: for some use cases, it may be useful to add a prev() method to go backwards in an array, or setInternalOffset(). The presence of gaps in arrays may make using it to implement algorithms such as binary search error-prone
- This may be useful to use with `ZEND_API zend_object_iterator *zend_user_it_get_new_iterator(zend_class_entry *ce, zval *object, int by_ref)` to allow user code to return a regular array in getIterator() and have php internally convert that to a zend_object_iterator for foreach to work without changing the many internal/external uses of `ce->get_iterator()`

morrisonlevi commented:
I think I should switch to `get_properties_for`, as you mention later. I don't like hiding the details (and it actually is more work, ha), but given reflection and other ways to manipulate PHP-level properties it can invalidate the state. I don't worry about such things generally speaking: if someone is breaking abstraction levels then they should accept the consequences if things go wrong. However, it _might_ be possible to cause a segfault, such as if we replace the inner array with some other type.

So yeah, I should hide the properties and use `get_properties_for`.

-----

I think partly you've mis-understood the intended design. Think of these as Array -> Iterator adapters. It takes an Array and packages it into an Iterator API for use in places where you specifically need an iterator (as I showed in the example in the original post). If we add array-specific features then it's leaking through the API. There may be exceptions to this -- for instance, a method like `toArray()` can be implemented more efficiently by the implementation than by using its public API, and it is a reasonable operation to perform on abstract iterators (hence `iterator_to_array`).

-----

There aren't "holes" in the iterator, by the way. The array iterator reports what the array has:
```php
$x = ['a','b','c']; unset($x[1]);

$iterator = new Spl\ForwardArrayIterator($x);
foreach ($iterator as $key => $current) {
    echo "[{$key} => {$current}]\n";
}
/* results in:
[0 => a]
[2 => c]
 */
```

morrisonlevi commented:
I did the following in recent commits:

- Switched to `get_properties_for`; currently only supports it for debug.
- Added (un)serialization deny handlers (cannot serialize them).
- Extracted some of the commonalities between Forward- and Reverse- ArrayIterators, though there is more that could be done.
- Added tests for ReverseArrayIterator.

TysonAndre commented on ext/spl/spl_array.stub.php:
```diff
@@ -246,3 +248,45 @@ public function hasChildren() {}
+    public function valid(): bool {}
+
+    /** @return mixed */
+    public function key() {}
```
nit: My preference for static analysis/IDEs/opcache would be to set the return type to `int|string`, which can be updated if arrays get changed to other key types such as object.

morrisonlevi commented on ext/spl/spl_array.c:
```diff
@@ -1839,9 +1846,538 @@ PHP_METHOD(ArrayObject, __debugInfo)
+static void ArrayIterator_free_obj(zend_object *object)
+{
+	ArrayIterator *iterator = ArrayIterator_from_obj(object);
+	if (iterator->ht != &zend_empty_array) {
```
This can be simplified, yes. I accidentally left this in from when I was debugging a memory issue. Good catch.

morrisonlevi commented on ext/spl/spl_array.c:
```diff
@@ -1839,9 +1846,538 @@ PHP_METHOD(ArrayObject, __debugInfo)
+	ForwardArrayIterator *iterator = ArrayIterator_from_obj(Z_OBJ_P(ZEND_THIS));
+	zval *current = ForwardArrayIterator_current(iterator);
+	if (EXPECTED(current)) {
+		ZVAL_COPY(return_value, current);
```
I am unsure on the deref. I'll think on it some more.

morrisonlevi replied:
I don't think a DEREF is needed -- at least as written I get `int(2)`. In the back of my head by-value is taken care of by the iteration API somewhere; I think the VM will do `zend_copy_to_variable`. However, as Joe Watkins told me, it is a cheap operation to perform _to be sure_. Maybe extensions will call this API and not doing the DEREF would be a problem. I've convinced myself you are right about the DEREF, I think.

I don't know much about indirect. @nikic , can you chime in there?

TysonAndre commented on ext/spl/spl_array.stub.php:
```diff
@@ -280,7 +280,7 @@ public function count(): int {}
 
-    /** @return mixed */
+    /** @return int|string */
     public function key() {}
```
I meant something else:

```suggestion
    public function key(): int|string {}
```

The arginfo.h files didn't change at all - union types are parsed from the signature, not comments - comments are more for unrepresentable types like `resource`

```
class ArrayIterator implements SeekableIterator, ArrayAccess, Serializable, Countable
{
    public function __construct(array|object $array = [], int $flags = 0) {}
```

morrisonlevi replied:
Ah, oops. I knew what you meant, I just did the wrong thing in a hurry.

Girgias commented:
I'm going to close this (thanks GH for pinging me...) I'm not against this feature per se, but I don't think it should use ``SPL`` as namespace. Also needs a rebase.

Please reopen, or open a new PR if this is still something that is being considered.
